# Population Size versus Number of Generations

Investigating the effects of decreasing the population sizes and increasing the number of generations (and vice versa).

## Hypothesis

We know that increasing the number of generations is likely to increase the fitness of our populations.
We also know that, if a population is too small, it will have difficulties maintaining itself.

So, let us hypothesize that we will get better players by increasing the iteration count and decreasing the population size than if we did the reverse.

## What we will be testing

Game: We will run our tests using Competitive Graph Coloring and our 14-node graph from a previous Notebook.

Parameters:
* Our fitness threshold will start at 0.5 and will not increase for the duration of our tests
* The number of games each player plays per generation will remain at 10.
* Mutation rate will remain at 0.025 for all tests.
* Our control group will be a population size of 100 on 10 iterations.
* Our bounds will be fixed at 3. We know this graph has a known game chromatic color of 4, and that it is *possible* to win with 3 colors.

So, the *only* parameters we are testing are population size and number of iterations.

## Setup

We need to import the necessary files and create a graph and ruleset.

In [1]:
import sys
sys.path.append("../../simple_games")

from classes.genetic_algorithm import GeneticAlgorithm

# Import the rulesets and strategy
from graph_coloring.classes.gc_ruleset import GCRuleset
from graph_coloring.classes.gc_random_init_strategy import GCRandomInitStrategy

In [2]:
# Create a graph to play on
initial_state = [
    {"color": 0, "adj": [5]},
    {"color": 0, "adj": [6]},
    {"color": 0, "adj": [7]},
    {"color": 0, "adj": [8]},
    {"color": 0, "adj": [5]},
    {"color": 0, "adj": [0,4,6,10]},
    {"color": 0, "adj": [1,5,7,11]},
    {"color": 0, "adj": [2,6,8,12]},
    {"color": 0, "adj": [3,7,9,13]},
    {"color": 0, "adj": [8]},
    {"color": 0, "adj": [5]},
    {"color": 0, "adj": [6]},
    {"color": 0, "adj": [7]},
    {"color": 0, "adj": [8]},
]

In [3]:
# Ruleset of the game being played on
ruleset = GCRuleset("Graph Coloring Ruleset", initial_state, bounds = 3)

Let's also define some functions to allow us run a test and see information about the player populations.

In [10]:
def print_pop(pop, fitness = 0.5):
    """
    Displays individuals in the population whose fitness is at or above the provided threshold.
    
    Args:
        pop : Population to display
        fitness (Optional) : Minimum fitness threshold
    """
    for p in pop:
        if p.fitness() >= fitness:
            print(p)

            
            
def percent_good_players(pop, fitness = 0.5):
    """
    Prints the percentage of a population that is above a specified fitness.
    
    Args:
        pop : Population to analyze
        fitness (Optional) : Minimum fitness threshold
    """
    good_pop = [p for p in pop if p.fitness() >= fitness]
    percent = len(good_pop) / len(pop)
    print("{:.2f}% above fitness {}".format(percent * 100, fitness))
    
    
    
def fitness_thresholds(p1_pop, p2_pop):
    """
    Displays the fitness thresholds for each population.
    
    Args:
        p1_pop : Player 1 population
        p2_pop : Player 2 population
    """
    print("Player 1 Fitness Thresholds:")
    for i in range(0, 11, 1):
        percent_good_players(p1_pop, fitness=i/10)

    print("\nPlayer 2 Fitness Thresholds:")
    for i in range(0, 11, 1):
        percent_good_players(p2_pop, fitness=i/10)

        
        
def run_test(pop_size, iterations):
    """
    Runs a series of genetic algorithms, averaging the results.
    
    Args:
        pop_size : Population size to test
        iterations : Number of iterations to test
    """
    total_p1_pop = []
    total_p2_pop = []
    
    # Run 10 tests
    for i in range(10):
        # Create a new Evolution instance with the example strategy
        test_group = GeneticAlgorithm(
            ruleset,
            # Random-on-initialization strategy for generating populations of random players
            random_on_init_strat = GCRandomInitStrategy,
            # Data to be used by the above strategy
            strat_data = {"vertices": range(len(ruleset.initial_state)), "colors": range(1, ruleset.bounds + 1)},
            # Size of the populations
            pop_size = pop_size,
            # Number of generations to iterate through
            iterations = iterations,
            # Minimum number of games each player must play during a generation
            num_games = 10,
            # Starting fitness threshold
            fitness = 0.5,
            # Maximum fitness threshold; be careful of setting this too close to 1.0
            max_fitness = 0.5,
            # How much the fitness threshold should increment after each iteration
            fitness_increment = 0.0,
            # Chance of a mutation to occur during player reproduction
            mutation_rate = 0.025
        )
    
        # Run the algorithm
        p1_pop, p2_pop = test_group.evolve()
        
        # Append to populations
        total_p1_pop.extend(p1_pop)
        total_p2_pop.extend(p2_pop)
    
    fitness_thresholds(total_p1_pop, total_p2_pop)

## Control Group

Our control group will be a population size of 100 with 10 iterations. As stated earlier, fitness and number of games will remain fixed.

In [5]:
run_test(pop_size=100, iterations=10)

Player 1 Fitness Thresholds:
100.00% above fitness 0.0
95.20% above fitness 0.1
95.20% above fitness 0.2
95.20% above fitness 0.3
95.20% above fitness 0.4
93.50% above fitness 0.5
83.40% above fitness 0.6
70.10% above fitness 0.7
56.60% above fitness 0.8
42.30% above fitness 0.9
24.00% above fitness 1.0

Player 2 Fitness Thresholds:
100.00% above fitness 0.0
6.60% above fitness 0.1
6.60% above fitness 0.2
6.60% above fitness 0.3
6.60% above fitness 0.4
6.60% above fitness 0.5
1.10% above fitness 0.6
0.00% above fitness 0.7
0.00% above fitness 0.8
0.00% above fitness 0.9
0.00% above fitness 1.0


The Player 1 population is doing very well. The Player 2 population is not doing nearly as well.
* Majority of P1 (56.6%) was at or above 0.8, 24% were perfect players.
* Only 1.1% of P2 was at or above 0.6, 6.6% were below, but all were non-zero.

## Test 1: Increase Iterations

Here we will examine the effects of increasing the number of iterations. Let's do a simple test of doubling them to 20.

In [6]:
run_test(pop_size=100, iterations=20)

Player 1 Fitness Thresholds:
100.00% above fitness 0.0
99.40% above fitness 0.1
99.40% above fitness 0.2
99.40% above fitness 0.3
99.40% above fitness 0.4
99.30% above fitness 0.5
93.50% above fitness 0.6
80.10% above fitness 0.7
64.40% above fitness 0.8
45.70% above fitness 0.9
20.70% above fitness 1.0

Player 2 Fitness Thresholds:
100.00% above fitness 0.0
1.40% above fitness 0.1
1.40% above fitness 0.2
1.40% above fitness 0.3
1.40% above fitness 0.4
1.40% above fitness 0.5
0.00% above fitness 0.6
0.00% above fitness 0.7
0.00% above fitness 0.8
0.00% above fitness 0.9
0.00% above fitness 1.0


There seems to be a slight improvement in the Player 1 population and slight decline in the Player 2 perfomance.
* P1 had improvements across the board except at perfect win rates, which declined to 20.7%
* All P2 was nonzero, but only 1.4% was at or above 0.5.

## Test 2: Further Increasing Iterations 

Let's increase the iterations to an arbitrarily high number, like 50.

In [7]:
run_test(pop_size=100, iterations=50)

Player 1 Fitness Thresholds:
100.00% above fitness 0.0
100.00% above fitness 0.1
100.00% above fitness 0.2
100.00% above fitness 0.3
100.00% above fitness 0.4
100.00% above fitness 0.5
96.20% above fitness 0.6
82.10% above fitness 0.7
63.30% above fitness 0.8
48.70% above fitness 0.9
12.60% above fitness 1.0

Player 2 Fitness Thresholds:
100.00% above fitness 0.0
3.50% above fitness 0.1
3.50% above fitness 0.2
3.50% above fitness 0.3
3.50% above fitness 0.4
3.50% above fitness 0.5
0.50% above fitness 0.6
0.10% above fitness 0.7
0.00% above fitness 0.8
0.00% above fitness 0.9
0.00% above fitness 1.0


A near total increase in performace from both Player 1 and Player 2 populations.
* All Player 1s had a fitness at or above 0.5, with nearly half (48.7%) at 0.9 and 12.6% being perfect players.
* All P2 were non-zero, 3.5% were above 0.5, and only 0.1% were above 0.7.

## Test 3: Increasing Population Size

Now we will turn our attention to changing the population sizes. Let's begin by doubling the population size from 100 to 200.

In [8]:
run_test(pop_size=200, iterations=10)

Player 1 Fitness Thresholds:
100.00% above fitness 0.0
93.50% above fitness 0.1
93.50% above fitness 0.2
93.50% above fitness 0.3
93.35% above fitness 0.4
91.25% above fitness 0.5
75.80% above fitness 0.6
59.05% above fitness 0.7
45.65% above fitness 0.8
35.45% above fitness 0.9
22.95% above fitness 1.0

Player 2 Fitness Thresholds:
100.00% above fitness 0.0
7.40% above fitness 0.1
7.40% above fitness 0.2
7.40% above fitness 0.3
7.40% above fitness 0.4
7.40% above fitness 0.5
0.80% above fitness 0.6
0.05% above fitness 0.7
0.00% above fitness 0.8
0.00% above fitness 0.9
0.00% above fitness 1.0


The results are almost identical to the control group. The Player 1 population performed slightly worse and the Player 2 population performed slightly better.
* Majority of P1 (59.85%) was at or above 0.7. 22.95% were perfect players.
* 7.4% of P2 was at or above 0.5, and 0.05% were above 0.7.

## Test 4: Further Increasing Population Size

Now we will increase our player size by an order of magnitude- from 100 to 1,000.

In [9]:
run_test(pop_size=1000, iterations=10)

Player 1 Fitness Thresholds:
100.00% above fitness 0.0
94.39% above fitness 0.1
94.39% above fitness 0.2
94.39% above fitness 0.3
94.27% above fitness 0.4
92.45% above fitness 0.5
71.61% above fitness 0.6
50.77% above fitness 0.7
37.50% above fitness 0.8
29.67% above fitness 0.9
18.66% above fitness 1.0

Player 2 Fitness Thresholds:
100.00% above fitness 0.0
4.75% above fitness 0.1
4.75% above fitness 0.2
4.75% above fitness 0.3
4.75% above fitness 0.4
4.75% above fitness 0.5
0.36% above fitness 0.6
0.04% above fitness 0.7
0.00% above fitness 0.8
0.00% above fitness 0.9
0.00% above fitness 1.0


These results are almost identical to the previous test's except that the P1 population has fewer players above 0.6.
* Majority of P1 (50.77%) was at or above 0.7, and 18.66% were perfect players. From 0.0 to 0.5 there were about 1% more than in Test 3.
* All P2 were non-zero. 4.75% were above 0.5, 0.04% were above 0.7.

## Conclusion

As the number of iterations increased:
* Player 1's tended to gain fitness overall.
* Fewer Player 1's were perfect (1.0 fitness) players.
* Player 2's tended to lose fitness overall.
* Slightly more Player 2's had higher fitnesses.

As the population size increased:
* A general decline in the fitness for Player 1
* The Player 2 population tends to remain about the same. No significant consistent change occurs.